oneflow/python/ops/nn_ops.py

```diff
@@ -2646,3 +2646,264 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
+
+        import oneflow as flow 
+        import oneflow.typing as tp 
+        import numpy as np 
+
+
+        @flow.global_function()
+        def triplet_loss_job(anchor: tp.Numpy.Placeholder(shape=(3, 3)),
+                            pos: tp.Numpy.Placeholder(shape=(3, 3)),
+                            neg: tp.Numpy.Placeholder(shape=(3, 3)))->tp.Numpy:
+            out = flow.nn.TripletMarginLoss(anchor, pos, neg, margin=1.0, p=2.0)
+            return out 
+
+        np_anchor = np.array([[1, 2, 3],
+                            [4, 5, 6],
+                            [7, 8, 9]]).astype(np.float32)
+
+        np_pos = np.array([[2, 2, 2],
+                        [2, 2, 2],
+                        [2, 2, 2]]).astype(np.float32)
+
+        np_neg = np.array([[3, 3, 3],
+                        [3, 3, 3],
+                        [3, 3, 3]]).astype(np.float32)
+
+        out = triplet_loss_job(np_anchor, np_pos, np_neg)
+
+        # output [1.8449262]
+
+    Args:
+        anchor (remote_blob_util.BlobDef): The anchor Blob. 
+        positive (remote_blob_util.BlobDef): The positive sample Blob. 
+        negative (remote_blob_util.BlobDef): The negative sample Blob. 
+        margin (float, optional): The margin value. Defaults to 1.0.
+        p (float, optional): The norm degree for computing distance. Defaults to 2.0.
+        eps (float, optional): A small value use in norm computation. Defaults to 1e-6.
+        swap (bool, optional): Whether to swap the distance. 
+        For more details you can check the Paper `Learning shallow convolutional feature descriptors with triplet losses`. Defaults to False.
+        reduction (str, optional): The reduce type, it can be one of "none", "mean", "sum". Defaults to "mean".
+        name (Optional[str], optional): The name for the operation. Defaults to None.
+
+    Returns:
+        remote_blob_util.BlobDef: The result Blob. 
+    """
+    assert reduction in [
+        "none",
+        "mean",
+        "sum",
+    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
+        reduction
+    )
+
+    assert (
+        swap == False
+    ), "For now we only support `swap=True`, OneFlow still have backward error in minimum"
+
+    if name is None:
+        name = "TripletMarginLoss_"
+
+    def _p_norm(x, p=2.0, name="p_norm"):
+        r"""Compute the p-norm 
+
+        The equation is: 
+
+        .. math:: 
+
+            out = \sqrt[P]{\sum_{i=0}^{n}(abs(x)^P)} 
+
+        Args:
+            x ([type]): The input Blob. 
+            p ([type], optional): The norm degree. Defaults to 2..
+        """
+        # In order to avoid the `nan` case.
+        _abs_val = flow.math.abs(x, name=name + "_abs")
+
+        if p == 2.0:
+            # Use Square to compute the p2-norm
+            _norm = flow.math.square(_abs_val, name=name + "_square")
+            _norm = flow.math.reduce_sum(_norm, axis=-1, name=name + "_sum")
```

Comment:
triplet loss 默认输入是只有两维的是吧，我看了下 pytorch 的代码，貌似reduce 的维度默认是在第一维，如果是两维的就是一样的。

Reply:
拿代码试验了下还真是，这个我改下

Reply:
先看到 `pairwise_distance`
```python3
Tensor pairwise_distance(const Tensor& x1, const Tensor& x2, double p, double eps, bool keepdim) {
  return at::norm(x1 - x2 + eps, p, 1, keepdim);
}
```
https://github.com/pytorch/pytorch/blob/25ba802ce4cbdeaebcad4a03cec8502f0de9b7b3/aten%2Fsrc%2FATen%2Fnative%2FDistance.cpp#L15

再跟进去看 `norm`，可以看到dim传的是1

```python3
static Tensor norm(const Tensor& self, optional<Scalar> p, IntArrayRef dim, bool keepdim,
            optional<ScalarType> opt_dtype) {
  if (self.is_sparse()) {
    // Sparse tensors need a different implementation because their values
    // are accessed with a different API than strided tensors
    return at::native_norm(self, p, dim, keepdim, opt_dtype);
  } else {
    Tensor result;
    return at::native::norm_out(result, self, p, dim, keepdim, opt_dtype);
  }
}
```
https://github.com/pytorch/pytorch/blob/ab0a04dc9c8b84d4a03412f1c21a6c4a2cefd36c/aten%2Fsrc%2FATen%2Fnative%2FReduceOps.cpp#L513